Signature algorithm extension parsing

yanesca opened the pull request:
## Description

Signature algorithm extension was skipped when renegotiation was in
progress, causing the signature algorithm not to be known when
renegotiating, and failing the handshake. Fix removes the renegotiation
step check before parsing the extension.

## Status
**READY**

## Requires Backporting
Backport PRs for 2.1 and 1.3 are ready.

## Todos
- [x] Tests
- [x] Changelog updated
- [x] Backported

Patater commented on tests/ssl-opt.sh:
```diff
@@ -1401,6 +1401,22 @@ run_test    "Renegotiation: client-initiated" \
 
+# Checks if the server parses the SigAlgs Extension. Valid only if an MD
+# algorithm stronger than SHA1 is enabled in config.h
+run_test    "Renegotiation: Signature Algorithms regression, client" \
```
"regression" is a name we will forget the meaning of easily. Please consider a better name.

Patater replied:
Might also be useful to say "client initiated" instead of just "client" (and likewise for server).

Patater commented on tests/ssl-opt.sh:
```diff
@@ -1401,6 +1401,22 @@ run_test    "Renegotiation: client-initiated" \
+            -c "found renegotiation extension" \
+            -c "=> renegotiate" \
+            -s "=> renegotiate" \
+            -S "client hello v3, signature_algorithm ext: 2" \
```
How obvious is it that "signature_algorithm ext: 2" is SHA-1? Does this warrant a comment?

Patater commented on tests/ssl-opt.sh:
```diff
@@ -1414,6 +1430,22 @@ run_test    "Renegotiation: server-initiated" \
             -s "=> renegotiate" \
             -s "write hello request"
 
+# Checks if the server parses the SigAlgs Extension. Valid only if an MD
```
Should this test be placed immediately after the client ("Renegotiation: Signature Algorithms regression, client") one? Why not?

yanesca replied:
I was grouping them to other renegotiation tests that are initiated by the same peer. But it seems that the tests around them are grouped by test type, so you are right that it may be better here to.

hanno-arm commented on tests/ssl-opt.sh:
```diff
@@ -1401,6 +1401,22 @@ run_test    "Renegotiation: client-initiated" \
             -S "write hello request"
 
+# Checks if the server parses the SigAlgs Extension. Valid only if an MD
+# algorithm stronger than SHA1 is enabled in config.h
```
We have `requires_config_enabled` for that purpose, which however only accepts only a single configuration flag.

hanno-arm replied:
@yanesca Could you extend the comment to explain a bit how the test is supposed to work? It took me a bit to understand.

hanno-arm replied:
We could use something like 
```
# skip next test if none of the flags is enabled in config.h
requires_config_enabled_or() {
    SKIP_NEXT="YES"
    for var in "$@"
    do
        if grep "^#define $var" $CONFIG_H > /dev/null; then
            SKIP_NEXT="NO"
        fi
    done
}
```
and prefix the test with `requires_config_enabled_or MBEDTLS_SHA256_C MBEDTLS_SHA512_C`.

yanesca replied:
I think this is a separate issue and bigger on its own than the current one. This isn't the first and only test where this feature would be necessary to detect compilation flags properly.

Fixing this problem should involve not just adding the feature to the script, but going through all the tests and flagging them properly.

In summary: I think this is out of scope and would prefer to leave it like it is.

hanno-arm replied:
@yanesca Fair enough. I'd still prefer to just add it for this test only so that it's easier in the future to adapt the other tests, but you're right that this critical PR should probably be kept as small as possible.

yanesca replied:
@hanno-arm It is not just because this PR is critical, in general I prefer if a PR (not just mine) only does what it needs to do and doesn't add a lot of ad hoc changes.

Patater replied:
Is there an issue tracking this problem generally? If not, could you please file one appropriately?

yanesca replied:
Yes, this issue is already being tracked internally.

hanno-arm commented on ChangeLog:
```diff
@@ -9,6 +9,9 @@ Bugfix
      Found by projectgus and jethrogb, #836.
    * Fix usage help in ssl_server2 example. Found and fixed by Bei Lin.
+   * Parse signature algorithm extension when renegotiating. Previously,
+     renegotiating handshakes would only accept signatures using SHA-1,
```
This could be expanded a little more to say "would only accept signatures using SHA-1 regardless of the client's preferences`.

yanesca commented:
@Patater @hanno-arm Thank you very much for reviewing it so expeditiously! I have rebased the branch and added commits to address the review feedback. Could you please review it again?

Patater commented on tests/ssl-opt.sh:
```diff
@@ -1428,11 +1428,13 @@ run_test    "Renegotiation: server-initiated" \
             -s "=> renegotiate" \
-            -s "write hello request"
+            -S "write hello request" \
+            -S "client hello v3, signature_algorithm ext: 2" # Is SHA-1 negotiated?
```
I think comments only work at the start of lines, based on what I see in `tests/suites/main_test.function`.

hanno-arm replied:
@Patater At least on my machine it appears to work.

hanno-arm replied:
@Patater `test/suites/main_test.function` is only for generating the test suite source files. It's not related to `ssl-opt.sh` to my knowledge.

yanesca replied:
Here comments work at the end of the line too, because`tests/ssl-opt.sh` is a plain bash script, whereas `tests/suites/main_test.function` is used by a test suite generation script to compile it to C source code.

Patater replied:
Yay, I like being able to have comments.

OK

yanesca commented:
@hanno-arm Thank you for your review!

@Patater I am going to backport this as soon as you approve!

gilles-peskine-arm commented:
Since this PR is based on the (current state of) #1123 , shouldn't 1123 be closed in favor of this one?

yanesca commented:
@gilles-peskine-arm Thank you very much for your review! Yes, I think it would be sensible to close #1123. Do you want me to do it?

yanesca commented:
Fixed the typos @hanno-arm found in the backports in this PR too.

sbutcher-arm commented:
Reviewed, tested, approved and merged into `development`.
